### PR TITLE
feat: support findByFilters

### PR DESCRIPTION
**Pull request by pivstone**
There are two approaches to implement this findByFilter:
- N1QL
- View query

### N1QL vs View Query

feature  | N1QL | ViewQuery
---|----|----
CouchBase3 | ✖️| ✔️
CouchBase4 | ✔️| ✔️
Where Basic Filter | ✔️ | ✔️
AND & OR operator |  ✔️ | ✖️
Range Query | ✔️ |  ✖️
inq | ✔️ | ✖️
near | ✖️ |  ✖️
not like | ✔️| ✖️
regexp | ✔️* | ✖️
Like Query | ✔️ | ✖️
Limit Filter | ✔️ | ✔️
Skip Filter |  ✔️ |  ✔️
Order Filter |  ✔️ | ✖️
Include Filter | Not  In schedule | Not  In the schedule 
Fields Filter | ✔️ | Not Sure
Node API | ✔️ | Partial 
REST API | ✔️| ✖️
SQL Inject | Safe * | Safe

### Notes:

- Regexp: The CouchBase use Golang to implement the pattern matching.  [Supported Syntax](https://github.com/google/re2/wiki/Syntax)
There are two examples:
```JS
 Person.find({ where: { name: { regexp: '^C+.*' } } });
```
```JS
 Person.find({ where: { name: { regexp: /^C+.*/ } } });
```

- SQL injection: The n1ql is generated via concatenating the string. But the parameters do not include in the query. The parameters will be escaped by CouchBase itself. For the reason, it's free from SQL injection.

### AND & OR operator

- Example
```JS
Person.find(
      { where: {
          and: [
            { age: { gt: 10 } },
            { age: { lt: 40 } }],
          or: [
            { name: 'Charlie' },
            { name: 'Mary' }]
        },
        order: 'name ASC' })
```

- NOT Support this kinda query:
```JS
Person.find({ where: { age: {  and: [ { gt: 10 }, { lt: 40 } ] } });
```
### How to enable the experimental feature:

Add the experiment field into `datasource.json`
```JS
{
    cluster: {
      url: process.env.COUCHBASE_URL || 'couchbase://localhost',
      options: {}
    },
    bucket: {
      name: 'test_bucket',
      password: '',
      operationTimeout: 60 * 1000
    },
    experiment: ['viewQuery'] // or n1ql
  };
```

### Add Dependency

- `esprima`: To find out all the call expression of `Model.find` via static code analysis

### CODE REVIEW

**Comment by pivstone** [comment]
Forget to update the current state:

- N1QL is fully supported in CoucBase4. However, it is an experimental feature in CoucBase3
- N1QL has a load imbalance issue.

For those reasons, I plan:

- To keep N1QL as an experimental feature, only support CoucBase4
- To implement the find via `View Query`, support CoucBase3 and CoucBase4 

Proposal:

- To remove the node 6 from the support list ( for the nodejs ecosystem)

BTW, `loopback-connector-couchbase` is a better one. We need to support the CouchBase5 in future.

**Comment by pivstone** [comment]
Update:

- Restructuring: the `N1QL` and `viewQuery` move to the `experiment feature`
- Add dynamic create view via static code analysis at `autoupdate` phase

How to enable the experimental feature:

- Add the experiment field into `datasource.json` 
```JS
{
    cluster: {
      url: process.env.COUCHBASE_URL || 'couchbase://localhost',
      options: {}
    },
    bucket: {
      name: 'test_bucket',
      password: '',
      operationTimeout: 60 * 1000
    },
    experiment: ['viewQuery']
  };
```

File: lib/utils.js [hunk 1]
```diff
@@ -37,4 +41,262 @@ function promisifier(originalMethod) {
...
+ *
+ * @param {Array} models The list of models
+ */
+exports.generateViews = function(models) {
+  var designDocs = { query: { views: { } } };
+  var methods = findAllModelMethod(models);
+  Object.keys(methods).forEach(key => {
+    var model = methods[key];
+    model.forEach(e => {
+      var fields = e.split(',').map(e => { return 'doc.' + e; }).join(',');
```
Comment (xavierchow) [inline lib/utils.js]: no validation for the `e`?

File: lib/experiment/n1ql.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
+'use strict';
+
+const Promise = require('bluebird');
+const debug = require('debug')('loopback:connector:couchbase3:n1ql');
+const N1qlQuery = require('couchbase').N1qlQuery;
+const utils = require('../utils');
+
+module.exports.Couchbase = {
+  /**
+   * Implement `autoupdate()`.
```
Comment (xavierchow) [inline lib/experiment/n1ql.js]: no need reimplement autoupdate, polymorphism would work.

File: lib/experiment/n1ql.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    debug('autoupdate', this.settings);
+    // Create views.
+    let designDocs = this.defaultDesignDocs;
+    if (this.settings.designDocs) {
+      designDocs = Object.assign({}, designDocs, this.settings.designDocs);
+    }
+    let promise = Promise.resolve(true);
+    for (let name in designDocs) {
+      promise = promise.then(() => {
+        return this.createPrimaryIndex({ ignoreIfExists: true }).then(() => {
```
[the remark this answers is not in the document]
Comment (pivstone) [reply lib/experiment/n1ql.js]: { ignoreIfExists: true }

**Comment by xavierchow** [comment]
no rush to fix it, we need a discussion later.

**Pull request closed without merging**